rts/motoko-rts/src/memory/ic.rs

```diff
@@ -0,0 +1,85 @@
+// This module is only enabled when compiling the RTS for IC or WASI.
+
+use super::Memory;
+use crate::rts_trap_with;
+use crate::types::*;
+
+use motoko_rts_macros::ic_fn;
+
+use core::arch::wasm32;
+
+/// Maximum live data retained in a GC.
+pub(crate) static mut MAX_LIVE: Bytes<u32> = Bytes(0);
+
+/// Amount of garbage collected so far.
+pub(crate) static mut RECLAIMED: Bytes<u64> = Bytes(0);
+
+/// Counter for total allocations
+pub(crate) static mut ALLOCATED: Bytes<u64> = Bytes(0);
+
+/// Heap pointer
+pub(crate) static mut HP: u32 = 0;
+
+// Provided by generated code
+extern "C" {
+    pub(crate) fn get_heap_base() -> u32;
+    pub(crate) fn get_static_roots() -> SkewedPtr;
+}
+
+#[ic_fn]
+unsafe fn init() {
+    HP = get_heap_base() as u32;
+}
+
+#[ic_fn]
+unsafe fn get_max_live_size() -> Bytes<u32> {
+    MAX_LIVE
+}
+
+#[ic_fn]
+unsafe fn get_reclaimed() -> Bytes<u64> {
+    RECLAIMED
+}
+
+#[ic_fn]
+unsafe fn get_total_allocations() -> Bytes<u64> {
+    ALLOCATED
+}
+
+#[ic_fn]
+unsafe fn get_heap_size() -> Bytes<u32> {
+    Bytes(HP - get_heap_base())
+}
+
+/// Provides a `Memory` implementation, to be used in functions compiled for IC or WASI. The
+/// `Memory` implementation allocates in Wasm heap with Wasm `memory.grow` instruction.
+pub struct IcMemory;
+
+impl Memory for IcMemory {
+    unsafe fn alloc_words(&mut self, n: Words<u32>) -> SkewedPtr {
+        let bytes = n.to_bytes();
+        // Update ALLOCATED
+        ALLOCATED += Bytes(bytes.0 as u64);
+
+        // Update heap pointer
+        let old_hp = HP;
+        let new_hp = old_hp + bytes.0;
+        HP = new_hp;
+
+        // Grow memory if needed
+        grow_memory(new_hp as usize);
+
+        skew(old_hp as usize)
+    }
+}
+
+unsafe fn grow_memory(ptr: usize) {
+    let total_pages_needed = ((ptr / 65536) + 1) as i32;
```

Comment:
Nit: Is there a named constant that can be used instead of 65536?

Reply:
This is Wasm page size (64KiB), I'll add a constant for this.

I'm a bit confused about what this code is doing though. I think it should divide by page size, round up. But the code is not doing that. It should be

```rust
(ptr + WASM_PAGE_SIZE - 1) / WASM_PAGE_SIZE
```

Current version returns a different result for e.g. `WASM_PAGE_SIZE`, it returns 2 but the correct value would be 1.

So I think it allocates more than necessary when `ptr` is a multiple of `WASM_PAGE_SIZE`.

@ulan do you agree with my analysis?

Reply:
To avoid overflows when `ptr` is larger than `u32::MAX - WASM_PAGE_SIZE` we should do this arithmetic on `u64`s.

Reply:
I agree with the analysis. The overflow handling is indeed subtle.
+1 to doing the expected arithmetic in `u64`

Reply:
I've added a constant for Wasm page size in #2632.

Reply:
`grow_memory` refactored in #2633. Marking as resolved.